### PR TITLE
feat(worker): support ExcelFileWriter action

### PR DESCRIPTION
**Pull request by pyshx**
# Overview
This PR adds support ExcelFileWriter Action in the `action-universal`.

## What I've done
I've implemented a `writer` action for `ActionValue::Map` and `ActionValue::Array`

Excel 2007+ xml formats (.xlsx)

We've implemented the following features:
1. Formatting(except Row Formatting)
2. ProtectionOptions
3. Formula
4. Hyperlinks

## What I haven't done
We haven't implemented the following features:
1. Row Formatting
2. Template File
3. Named Ranges
4. Pivot Tables, Charts and Graphs

## How I tested
I wrote a couple of tests along with the writer and used `cargo test`

## Which point I want you to review particularly
Please check if the assumed format for expected row data is correct and let me if we need to support the unsupported feature soon 🙏🏼

### CODE REVIEW

File: worker/crates/action-universal/src/excel/writer.rs [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+        "CenterAcross" => Ok(FormatAlign::CenterAcross),
+        "Distributed" => Ok(FormatAlign::Distributed),
+        "Top" => Ok(FormatAlign::Top),
+        "Bottom" => Ok(FormatAlign::Bottom),
+        "VerticalCenter" => Ok(FormatAlign::VerticalCenter),
+        "VerticalJustify" => Ok(FormatAlign::VerticalJustify),
+        "VerticalDistributed" => Ok(FormatAlign::VerticalDistributed),
+        _ => Err(error::Error::internal_runtime("Invalid alignment value")),
+    }
+}
```
Comment (rot1024) [inline worker/crates/action-universal/src/excel/writer.rs:326]: TryFrom can be used
```rust
impl TryFrom<&str> for FormatAlign {
    type Error = /* */;

    fn try_from(value: i32) -> Result<Self, Self::Error> {
		// ...
	}
}
```

Comment (pyshx) [reply worker/crates/action-universal/src/excel/writer.rs:326]: I thought of doing that but `FormatUnderline` or `FormatAlign` is not defined in the current crate but part of https://docs.rs/rust_xlsxwriter/0.64.1/rust_xlsxwriter/index.html

File: worker/crates/action-universal/src/excel/writer.rs [hunk 2]
```diff
@@ -0,0 +1,640 @@
...
+    _key: &str,
+    value: &ActionValue,
+) -> Result<()> {
+    let cell_value = match value {
+        ActionValue::String(s) => s.clone(),
+        ActionValue::Number(n) => n.to_string(),
+        _ => "".to_string(),
+    };
+
+    let _ = worksheet.write_string_with_format(
```
Comment (miseyu) [inline worker/crates/action-universal/src/excel/writer.rs:142]: > [!CAUTION]
> Error handling should be done.

File: worker/crates/action-universal/src/excel/writer.rs [hunk 3]
```diff
@@ -0,0 +1,640 @@
...
+    worksheet: &mut Worksheet,
+    row: usize,
+    col: usize,
+    key: &str,
+    row_data: &HashMap<String, ActionValue>,
+) -> Result<()> {
+    if let Some(ActionValue::String(formatting_str)) = row_data.get(&format!("{}.formatting", key))
+    {
+        let format = parse_formatting(formatting_str)?;
+        let _ = worksheet.write_string_with_format(row as u32, col as u16, "", &format);
```
Comment (miseyu) [inline worker/crates/action-universal/src/excel/writer.rs:162]: > [!CAUTION]
> Error handling should be done.

File: worker/crates/action-universal/src/excel/writer.rs [hunk 4]
```diff
@@ -0,0 +1,640 @@
...
+
+fn write_cell_formula(
+    worksheet: &mut Worksheet,
+    row: usize,
+    col: usize,
+    key: &str,
+    row_data: &HashMap<String, ActionValue>,
+) -> Result<()> {
+    if let Some(ActionValue::String(formula_str)) = row_data.get(&format!("{}.formula", key)) {
+        let _ = worksheet.write_formula(row as u32, col as u16, Formula::new(formula_str));
```
Comment (miseyu) [inline worker/crates/action-universal/src/excel/writer.rs:176]: > [!CAUTION]
> Error handling should be done.

File: worker/crates/action-universal/src/excel/writer.rs [hunk 5]
```diff
@@ -0,0 +1,640 @@
...
+
+fn write_cell_hyperlink(
+    worksheet: &mut Worksheet,
+    row: usize,
+    col: usize,
+    key: &str,
+    row_data: &HashMap<String, ActionValue>,
+) -> Result<()> {
+    if let Some(ActionValue::String(hyperlink_str)) = row_data.get(&format!("{}.hyperlink", key)) {
+        let _ = worksheet.write_url(row as u32, col as u16, Url::new(hyperlink_str));
```
Comment (miseyu) [inline worker/crates/action-universal/src/excel/writer.rs:190]: > [!CAUTION]
> Error handling should be done.

File: worker/crates/action-universal/src/excel/writer.rs [hunk 6]
```diff
@@ -0,0 +1,640 @@
...
+    Ok(())
+}
+
+fn write_map_entry(
+    worksheet: &mut Worksheet,
+    row_index: &mut usize,
+    key: String,
+    value: ActionValue,
+) -> Result<()> {
+    let _ = worksheet.write_string_with_format(*row_index as u32, 0, &key, &Default::default());
```
Comment (miseyu) [inline worker/crates/action-universal/src/excel/writer.rs:202]: > [!CAUTION]
> Error handling should be done.

File: worker/crates/action-universal/src/excel/writer.rs [hunk 7]
```diff
@@ -0,0 +1,640 @@
...
+
+    match value {
+        ActionValue::String(s) => {
+            let _ =
```
Comment (miseyu) [inline worker/crates/action-universal/src/excel/writer.rs:206]: > [!CAUTION]
> Error handling should be done.

File: worker/crates/action-universal/src/excel/writer.rs [hunk 8]
```diff
@@ -0,0 +1,640 @@
...
+                worksheet.write_string_with_format(*row_index as u32, 1, &s, &Default::default());
+        }
+        ActionValue::Number(n) => {
+            if let Some(num) = n.as_f64() {
+                let _ = worksheet.write_number(*row_index as u32, 1, num);
```
Comment (miseyu) [inline worker/crates/action-universal/src/excel/writer.rs:211]: > [!CAUTION]
> Error handling should be done.

File: worker/crates/action-universal/src/excel/writer.rs [hunk 9]
```diff
@@ -0,0 +1,640 @@
...
+            } else {
+                let _ = worksheet.write_string_with_format(
```
Comment (miseyu) [inline worker/crates/action-universal/src/excel/writer.rs:213]: > [!CAUTION]
> Error handling should be done.

File: worker/crates/action-universal/src/excel/writer.rs [hunk 10]
```diff
@@ -0,0 +1,640 @@
...
+                    *row_index as u32,
+                    1,
+                    &n.to_string(),
+                    &Default::default(),
+                );
+            }
+        }
+        ActionValue::Bool(b) => {
+            let _ = worksheet.write_boolean(*row_index as u32, 1, b);
```
Comment (miseyu) [inline worker/crates/action-universal/src/excel/writer.rs:222]: > [!CAUTION]
> Error handling should be done.

File: worker/crates/action-universal/src/excel/writer.rs [hunk 11]
```diff
@@ -0,0 +1,640 @@
...
+        }
+        ActionValue::Array(arr) => {
+            for (col_num, value) in arr.iter().enumerate() {
+                match value {
+                    ActionValue::String(s) => {
+                        let _ = worksheet.write_string_with_format(
```
Comment (miseyu) [inline worker/crates/action-universal/src/excel/writer.rs:228]: > [!CAUTION]
> Error handling should be done.

File: worker/crates/action-universal/src/excel/writer.rs [hunk 12]
```diff
@@ -0,0 +1,640 @@
...
+                            *row_index as u32,
+                            col_num as u16 + 1,
+                            s,
+                            &Default::default(),
+                        );
+                    }
+                    ActionValue::Number(n) => {
+                        if let Some(num) = n.as_f64() {
+                            let _ =
```
Comment (miseyu) [inline worker/crates/action-universal/src/excel/writer.rs:237]: > [!CAUTION]
> Error handling should be done.

File: worker/crates/action-universal/src/excel/writer.rs [hunk 13]
```diff
@@ -0,0 +1,640 @@
...
+                                worksheet.write_number(*row_index as u32, col_num as u16 + 1, num);
+                        } else {
+                            let _ = worksheet.write_string_with_format(
```
Comment (miseyu) [inline worker/crates/action-universal/src/excel/writer.rs:240]: > [!CAUTION]
> Error handling should be done.

File: worker/crates/action-universal/src/excel/writer.rs [hunk 14]
```diff
@@ -0,0 +1,640 @@
...
+                                *row_index as u32,
+                                col_num as u16 + 1,
+                                &n.to_string(),
+                                &Default::default(),
+                            );
+                        }
+                    }
+                    ActionValue::Bool(b) => {
+                        let _ = worksheet.write_boolean(*row_index as u32, col_num as u16 + 1, *b);
```
Comment (miseyu) [inline worker/crates/action-universal/src/excel/writer.rs:249]: > [!CAUTION]
> Error handling should be done.

File: worker/crates/action-universal/src/excel/writer.rs [hunk 15]
```diff
@@ -0,0 +1,640 @@
...
+
+async fn write_excel(
+    inputs: Option<ActionDataframe>,
+    props: &ExcelFileWriter,
+    storage_resolver: Arc<StorageResolver>,
+) -> Result<ActionValue> {
+    let mut workbook = Workbook::new();
+    let worksheet = workbook.add_worksheet();
+
+    let _ = worksheet.set_name(props.worksheet_name.clone());
```
Comment (miseyu) [inline worker/crates/action-universal/src/excel/writer.rs:53]: > [!CAUTION]
> Error handling should be done.

**Comment by miseyu** [comment]
@pyshx 
Overall, error handling is not done well, so please use question mark operator and map_err for better error handling. 🙇

**Review by miseyu — APPROVED** [review]
Thank you so much!!
LGTM

File: worker/crates/action-universal/src/excel/writer.rs (hunk 1 above)
Comment (miseyu) [reply worker/crates/action-universal/src/excel/writer.rs]: @pyshx 
I think we can implement TryFrom to FormatAlign for this.
Within this writer.rs file. The best thing to do would be to create a wrapper for Excel in common crate and implement TryFrom there as well, but I think we're good there.

Comment (pyshx) [reply worker/crates/action-universal/src/excel/writer.rs]: Understood! Can you check this https://github.com/reearth/reearth-flow/pull/138/commits/70fa5eafbe4b42048480414535b0ec200c1ec3a7 🙏🏼

File: worker/crates/action-universal/src/excel/writer.rs [hunk 16]
```diff
@@ -0,0 +1,681 @@
...
+//         let value = parts.next().ok_or_else(|| Error::internal_runtime("Invalid row formatting value"))?;
+//         match key {
+//             "row_height" => format.set_row_height(value.parse().map_err(|_| Error::internal_runtime("Invalid row height"))?),
+//             _ => return Err(Error::internal_runtime("Unknown row formatting key")),
+//         }
+//     }
+//     Ok(format)
+// }
+
+trait ExcelFormatParsable: Sized {
```
Comment (miseyu) [inline worker/crates/action-universal/src/excel/writer.rs:399]: I don't need this one.

File: worker/crates/action-universal/src/excel/writer.rs [hunk 17]
```diff
@@ -0,0 +1,681 @@
...
+    fn from_str(value: &str) -> Result<Self, String>;
+}
+
+struct ExcelFormatAlign(FormatAlign);
+
+impl ExcelFormatParsable for ExcelFormatAlign {
```
Comment (miseyu) [inline worker/crates/action-universal/src/excel/writer.rs:405]:
```suggestion
impl TryFrom<&str> for ExcelFormatAlign {
    type Error = Error;

    fn try_from(value: &str) -> Result<Self> {
        match value {
            "General" => Ok(ExcelFormatAlign(FormatAlign::General)),
            "Left" => Ok(ExcelFormatAlign(FormatAlign::Left)),
            "Center" => Ok(ExcelFormatAlign(FormatAlign::Center)),
            "Right" => Ok(ExcelFormatAlign(FormatAlign::Right)),
            "Fill" => Ok(ExcelFormatAlign(FormatAlign::Fill)),
            "Justify" => Ok(ExcelFormatAlign(FormatAlign::Justify)),
            "CenterAcross" => Ok(ExcelFormatAlign(FormatAlign::CenterAcross)),
            "Distributed" => Ok(ExcelFormatAlign(FormatAlign::Distributed)),
            "Top" => Ok(ExcelFormatAlign(FormatAlign::Top)),
            "Bottom" => Ok(ExcelFormatAlign(FormatAlign::Bottom)),
            "VerticalCenter" => Ok(ExcelFormatAlign(FormatAlign::VerticalCenter)),
            "VerticalJustify" => Ok(ExcelFormatAlign(FormatAlign::VerticalJustify)),
            "VerticalDistributed" => Ok(ExcelFormatAlign(FormatAlign::VerticalDistributed)),
            _ => Err(Error::internal_runtime(format!(
                "Invalid align value: {}",
                value
            ))),
        }
    }
}
```

File: worker/crates/action-universal/src/excel/writer.rs (hunk 16 above)
Comment (miseyu) [inline worker/crates/action-universal/src/excel/writer.rs]:
```suggestion
```

File: worker/crates/action-universal/src/excel/writer.rs (hunk 17 above)
Comment (miseyu) [inline worker/crates/action-universal/src/excel/writer.rs:404]:
```suggestion
impl TryFrom<&str> for ExcelFormatUnderline {
```

File: worker/crates/action-universal/src/excel/writer.rs [hunk 18]
```diff
@@ -0,0 +1,681 @@
...
+    fn from_str(value: &str) -> Result<Self, String> {
```
Comment (miseyu) [inline worker/crates/action-universal/src/excel/writer.rs]:
```suggestion
    type Error = Error;

    fn try_from(value: &str) -> Result<Self> {
```

File: worker/crates/action-universal/src/excel/writer.rs [hunk 19]
```diff
@@ -0,0 +1,681 @@
...
+            "VerticalJustify" => Ok(ExcelFormatAlign(FormatAlign::VerticalJustify)),
+            "VerticalDistributed" => Ok(ExcelFormatAlign(FormatAlign::VerticalDistributed)),
+            _ => Err(format!("Invalid alignment value: {}", value)),
+        }
+    }
+}
+
+struct ExcelFormatUnderline(FormatUnderline);
+
+impl ExcelFormatParsable for ExcelFormatUnderline {
```
Comment (miseyu) [inline worker/crates/action-universal/src/excel/writer.rs:430]:
```suggestion
impl TryFrom<&str> for ExcelFormatUnderline {
```

File: worker/crates/action-universal/src/excel/writer.rs [hunk 20]
```diff
@@ -0,0 +1,681 @@
...
+    fn from_str(value: &str) -> Result<Self, String> {
+        match value {
+            "None" => Ok(ExcelFormatUnderline(FormatUnderline::None)),
+            "Single" => Ok(ExcelFormatUnderline(FormatUnderline::Single)),
+            "Double" => Ok(ExcelFormatUnderline(FormatUnderline::Double)),
+            "SingleAccounting" => Ok(ExcelFormatUnderline(FormatUnderline::SingleAccounting)),
+            "DoubleAccounting" => Ok(ExcelFormatUnderline(FormatUnderline::DoubleAccounting)),
+            _ => Err(format!("Invalid underline value: {}", value)),
```
Comment (miseyu) [inline worker/crates/action-universal/src/excel/writer.rs]:
```suggestion
            _ => Err(Error::internal_runtime(format!(
                "Invalid underline value: {}",
                value
            ))),
```

File: worker/crates/action-universal/src/excel/writer.rs (hunk 1 above)
Comment (miseyu) [reply worker/crates/action-universal/src/excel/writer.rs]: @pyshx 
Sorry, I had to comment. 🙇

**Pull request merged**